src/node/controller/messages.rs

```diff
@@ -136,9 +137,53 @@ impl Controller {
     pub async fn handle_publish(
         &mut self,
         _header: &Header,
-        _publish: Publish,
+        publish: Publish,
     ) -> anyhow::Result<()> {
         // dbg!(publish);
+        let _block = match publish.block_holder {
+            BlockHolder::Send(_) => {
+                todo!("Received a send block")
+            }
+            BlockHolder::Receive(_) => {
+                todo!("Received a receive block")
+            }
+            BlockHolder::Open(_) => {
+                todo!("Received an open block")
+            }
+            BlockHolder::Change(_) => {
+                todo!("Received a change block")
+            }
+            BlockHolder::State(state_block) => {
+                let mut block = Block::from_state_block(&state_block);
+                let block_hash: &BlockHash = {
+                    block.calc_hash().unwrap();
+                    block.hash().unwrap()
+                };
+                // dbg!(state_block);
+
+                // # deduplication
+                let already_exists = self.state.lock().await
+                    .get_block_by_hash(block_hash).await?.is_some();
+
+                // # signature validation
+                let invalid_signature = || {
+                    let valid_signature = block.verify_self_signature()
```

Comment:
I think you can just do `block.verify_signature().is_ok()` or `.is_err()` to get a `bool`.

Does it need to be a closure?

Reply:
yes because otherwise it will be computed even if not necessary. On the other hand we could just use a different pattern. Remember the `if else` question in the discord? It was about this.

Reply:
I guess if it turns into the `is_ok()` variant, it won't be computed if the first `if` branch is true.

As in `} else if block.verify_self_signature().is_err() {`

Reply:
Sure, and for this example is fine because it can be reasonably turned into a one liner. I'm concerned whether it can be done for all. For example if the other checks are more elaborate they might not fit inline. I'll write a few more and see what's the result.

Reply:
I guess in that case the whole if statement could go into a function with early exits, similar to your initial example in chat?

Reply:
I don't think my first example would help. For reference it was:
```
if cond1 {
  do_bla();
  return Ok()
}
if cond2 {
  do_bla2();
  return Ok()
}
...
```
Now, if cond1 needs to lock the state and search the database for example, is not going to look pretty. This is why closures are a good way to group and name the checks on top of the `if`s, whatever pattern we chose. I'm trying to think of other options but basically what we have to implement here is a very extensive flow diagram. Any other idea?

Reply:
Well let's just leave it at what you think is best. If we can find a nicer solution later, it shouldn't be a big deal to change.